File: 3763-separate-squares-i/separate-squares-i.java
class Solution {
    public double separateSquares(int[][] squares) {
        double totalArea = 0;
        double low = Double.MAX_VALUE;
        double high = Double.MIN_VALUE;

        for (int[] s : squares) {
            int y = s[1];
            int l = s[2];
            totalArea += (double) l * l;
            low = Math.min(low, y);
            high = Math.max(high, y + l);
        }

        double halfArea = totalArea / 2.0;

        for (int i = 0; i < 60; i++) {
            double mid = (low + high) / 2.0;
            double areaBelow = 0;

            for (int[] s : squares) {
                int y = s[1];
                int l = s[2];
                if (mid <= y) continue;
                if (mid >= y + l) areaBelow += (double) l * l;
                else areaBelow += (double) l * (mid - y);
            }

            if (areaBelow < halfArea) low = mid;
            else high = mid;
        }

        return low;
    }
}

File: 3763-separate-squares-i/separate-squares-i.py
from typing import List

class Solution:
    def separateSquares(self, squares: List[List[int]]) -> float:
        total_area = sum(l * l for _, _, l in squares)
        half_area = total_area / 2.0

        low = min(y for _, y, _ in squares)
        high = max(y + l for _, y, l in squares)

        def area_below(h):
            area = 0.0
            for _, y, l in squares:
                if h <= y:
                    continue
                elif h >= y + l:
                    area += l * l
                else:
                    area += l * (h - y)
            return area

        for _ in range(60):
            mid = (low + high) / 2.0
            if area_below(mid) < half_area:
                low = mid
            else:
                high = mid

        return low
